fendermint/vm/interpreter/src/chain.rs

```diff
@@ -177,32 +217,87 @@ where
                     }
 
                     // We can use the same result type for now, it's isomorphic.
-                    Ok(((pool, state), ChainMessageApplyRet::Signed(ret)))
+                    Ok(((pool, provider, state), ChainMessageApplyRet::Signed(ret)))
                 }
                 IpcMessage::BottomUpExec(_) => {
                     todo!("#197: implement BottomUp checkpoint execution")
                 }
-                IpcMessage::TopDown => {
-                    todo!("implement TopDown handling; this is just a placeholder")
+                IpcMessage::TopDownExec(p) => {
+                    if !provider.is_enabled() {
+                        return Err(anyhow!(
+                            "cannot execute IPC top-down message: parent provider disabled"
+                        ));
+                    }
+
+                    // commit parent finality first
+                    let finality = IPCParentFinality {
+                        height: p.height as u64,
+                        block_hash: p.block_hash,
+                    };
+                    let msg = encode_to_fvm_implicit(&encode_commit_parent_finality_call(
+                        finality.clone(),
+                    )?)?;
+                    let (state, ret) = self
+                        .inner
+                        .deliver(state, VerifiableMessage::NotVerify(msg))
+                        .await?;
+                    if ret.is_err() {
+                        // TODO: how to handle error here? Do we return Err or stash them into one ret?
```

Comment:
Ah, wait, we are using the terms too loosely here. What do we mean by "return" exactly? 

My initial thought we can just  `return Err(anyhow!("failed to apply finality"))` and by doing that crash Fendermint, because if a legitimately voted upon finality fails to be added to the contract, we have a serious bug in the system. 

But you might argue it's better to keep chugging along, effectively severed from the parent subnet, in which case we can `return Ok(ret)` and go into terra incognita, see if we end up with a consensus failure or live to fight another day. 

Whichever it is, I don't think we should continue with the execution of messages.